src/scheduler.cc

```diff
@@ -14,35 +14,45 @@ namespace detail {
 	void abstract_scheduler::shutdown() { notify(scheduler_event_type::SHUTDOWN, 0); }
 
 	bool abstract_scheduler::is_idle() const noexcept {
-		std::lock_guard<std::mutex> lock(events_mutex);
-		return events.empty();
+		std::lock_guard lock(events_mutex);
+		return available_events.empty() && in_flight_count == 0;
 	}
 
 	void abstract_scheduler::schedule() {
-		std::unique_lock<std::mutex> lk(events_mutex);
-
 		while(true) {
-			// TODO: We currently operate in lockstep with the main thread. This is less than ideal.
-			events_cv.wait(lk, [this] { return !events.empty(); });
-			const auto event = events.front();
-			events.pop();
-
-			const task_id tid = event.data;
-			if(event.type == scheduler_event_type::TASK_AVAILABLE) {
-				naive_split_transformer naive_split(num_nodes, num_nodes);
-				ggen.build_task(tid, {&naive_split});
-				gsrlzr.flush(tid);
-			} else if(event.type == scheduler_event_type::SHUTDOWN) {
-				assert(events.empty());
-				return;
+			{
+				std::unique_lock lk(events_mutex);
+				events_cv.wait(lk, [this] { return !available_events.empty(); });
+				std::swap(available_events, in_flight_events);
+				in_flight_count = in_flight_events.size();
+			}
+
+			while(!in_flight_events.empty()) {
+				const auto event = std::move(in_flight_events.front());
```

Comment:
Clang-Tidy: `std::move` of the expression of the trivially-copyable type `__gnu_cxx::__alloc_traits<std::allocator<celerity::detail::scheduler_event>, celerity::detail::scheduler_event>::value_type` (aka `celerity::detail::scheduler_event`) has no effect; remove `std::move()`

Reply:
I got the same diagnostic and deliberately kept this so that if `scheduler_event` would for some reason no longer be trivially-copyable in the future, we're already set. As it says, it's a no-op right now so it doesn't really do any harm either, IMO.

Reply:
Fine with me, but we could still document the intent with a `// NOLINT(performance-move-const-arg)`.